bulldozer/merge.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2018 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bulldozer
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/github"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+
+	"github.com/palantir/bulldozer/pull"
+)
+
+const MaxPullRequestPollCount = 5
+
+func MergePR(ctx context.Context, pullCtx pull.Context, client *github.Client, mergeConfig MergeConfig) error {
+	logger := zerolog.Ctx(ctx)
+
+	mergeOpts := &github.PullRequestOptions{}
+
+	switch mergeConfig.Method {
+	case SquashAndMerge, MergeCommit, RebaseAndMerge:
+		mergeOpts.MergeMethod = string(mergeConfig.Method)
+	default:
+		mergeOpts.MergeMethod = string(MergeCommit)
+	}
+
+	commitMessage := ""
+	if mergeConfig.Method == SquashAndMerge {
+		opt, ok := mergeConfig.Options[SquashAndMerge]
+		if !ok {
+			logger.Error().Msgf("Unable to find matching %s in merge option configuration; using default %s", SquashAndMerge, EmptyBody)
+			opt = MergeOption{EmptyBody}
+		}
+
+		switch opt.Body {
+		case PullRequestBody:
+			body, err := pullCtx.Body(ctx)
+			if err != nil {
+				return errors.Wrap(err, "failed to determine pull request body")
+			}
+			commitMessage = body
+		case SummarizeCommits:
+			summarizedMessages, err := summarizeCommitMessages(ctx, pullCtx, client)
+			if err != nil {
+				return errors.Wrap(err, "failed to collect pull request commit messages")
+			}
+
+			commitMessage = summarizedMessages
+		case EmptyBody:
+		default:
+		}
+	}
+
+	go func(ctx context.Context) {
```

Comment:
style: given the complexity of this goroutine, it might be nice to break it out into a top-level private function, if it doesn't use too many references from the outer scope